bindings/go/constructor/construct.go

```diff
@@ -0,0 +1,499 @@
+	}
+
+	var creds map[string]string
+	if c.opts.CredentialProvider != nil {
+		identity, err := method.GetCredentialConsumerIdentity(ctx, src)
+		if err != nil {
+			return nil, fmt.Errorf("error getting credential consumer identity of type %q: %w", src.Input.GetType(), err)
+		}
+
+		if creds, err = c.opts.Resolve(ctx, identity); err != nil {
+			return nil, fmt.Errorf("error resolving credentials for input method of type %q: %w", src.Input.GetType(), err)
+		}
+	}
+
+	result, err := method.ProcessSource(ctx, src, creds)
+	if err != nil {
+		return nil, fmt.Errorf("error getting blob from input method: %w", err)
+	}
+
+	var processedSource *descriptor.Source
+
+	if result.ProcessedBlobData != nil {
+		processedSource, err = addColocatedSourceLocalBlob(ctx, targetRepo, component, version, src, result.ProcessedBlobData)
+	} else if result.ProcessedSource != nil {
+		processedSource = result.ProcessedSource
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error adding source %q to target repository: %w", src.ToIdentity(), err)
+	}
+	if processedSource == nil {
+		return nil, fmt.Errorf("input method for source %q did not return a processed source or blob data", src.ToIdentity())
+	}
+
+	return processedSource, nil
+}
+
+// processResourceWithInput handles the specific case of processing a resource that has an input method.
+// It looks up the appropriate input method from the registry and processes the resource
+// using the found method.
+func (c *DefaultConstructor) processResourceWithInput(ctx context.Context, targetRepo TargetRepository, resource *constructorv1.Resource, component, version string) (*descriptor.Resource, error) {
+	method, err := c.opts.GetResourceInputMethod(ctx, resource)
+	if err != nil {
+		return nil, fmt.Errorf("no input method resolvable for input specification of type %q: %w", resource.Input.GetType(), err)
+	}
+
+	var creds map[string]string
+	if c.opts.CredentialProvider != nil {
+		identity, err := method.GetCredentialConsumerIdentity(ctx, resource)
+		if err != nil {
+			return nil, fmt.Errorf("error getting credential consumer identity of type %q: %w", resource.Input.GetType(), err)
+		}
+
+		if creds, err = c.opts.Resolve(ctx, identity); err != nil {
+			return nil, fmt.Errorf("error resolving credentials for input method of type %q: %w", resource.Input.GetType(), err)
+		}
+	}
+
+	result, err := method.ProcessResource(ctx, resource, creds)
+	if err != nil {
+		return nil, fmt.Errorf("error getting blob from input method: %w", err)
+	}
+
+	var processedResource *descriptor.Resource
+
+	if result.ProcessedBlobData != nil {
+		processedResource, err = addColocatedResourceLocalBlob(ctx, targetRepo, component, version, resource, result.ProcessedBlobData)
+	} else if result.ProcessedResource != nil {
+		processedResource = result.ProcessedResource
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error adding resource %q to target repository: %w", resource.ToIdentity(), err)
+	}
```

Comment:
Why is this error checked down here and not where the error is set/created?

Reply:
To avoid nesting. IMO its always preferrable to have codepaths the least nested way possible. In this instance I could avoid having to write an if inside another if, because down the line this is an antipattern for large code flows. (similar as the other code path where you commented). Happy to hear your thoughts on this